### ISSUE TITLE
Can't compare KnownHostsFile instances

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6255]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6255|
|Type|defect|
|Created|2013-01-13 22:42:04Z|
|Branch|https://github.com/twisted/twisted/tree/knownhostsfile-savepath-6255|

This makes it difficult to test code using `KnownHostsFile`, for example to determine if the code has opened the right file.

The only identifying attribute of instances of this class is private, `_savePath`.  Making it public would be sufficient for testing purposes, since then code could at least just directly compare the path. The attribute should be made public in a read-only manner, however, since current code does not support changing it.

Attachments:

* [ticket6255.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/09a/09a5242c3c99d0c5291a9aaae5892412f36eaa3f/014ecfec798a845c61ffa5a294824eb271c838e0.patch) (421 bytes) - added by racheesingh on 2013-04-15 01:31:41Z - Patch file for the proposed fix for Ticket #6255
* [ticket6255-round2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/09a/09a5242c3c99d0c5291a9aaae5892412f36eaa3f/263a00fb68c17a608bbcbce28e0522fd0a63391d.patch) (2725 bytes) - added by racheesingh on 2013-04-16 01:57:02Z - Patch file for the proposed fix for Ticket #6255, incorporating review comments.

<details><summary>Searchable metadata</summary>

```
trac-id__6255 6255
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_knownhostsfile_savepath_6255 branches/knownhostsfile-savepath-6255
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__easy easy
time__1358116924000000 1358116924000000
changetime__1369915889000000 1369915889000000
version__None None
owner__exarkun exarkun
cc__z3p cc__racheesingh
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

There's an "or" in the ticket description.  For this ticket to actually be easy, that should probably be removed.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

I updated the ticket's description. 

For the record, I removed the following alternate suggestion, on the basis that comparison of contents might be another thing to compare, and it's not clear which one is the natural one for built-in operators:

```
Alternatively `KnownHostsFile` could support `==` and `!=` directly (via `__eq__` and `__ne__` - or better, `FancyEqMixin`).
```

**Comment by twisted-trac** [comment]
|<img alt="racheesingh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|racheesingh set owner to racheesingh<br>racheesingh set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="racheesingh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|racheesingh commented|
|-|-|

As per what I understood from the ticket description and comments, the preferable thing to do here is:
1. Make _savePath public.
2. Make _savePath read-only so no one tries to modify it (mistakenly).

Does that sound correct?

Should the patch include a test case to ensure that the proposed variable "savePath" (no longer private) cannot be written to (we could expect an Exception in the test case)?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

Those two are the requirements, not necessarily the implementation. You could achieve both by having a function that returns `_savePath`.

Any expected behavior should be tested. If you're writing a function that returns `_savePath` there's no public API to write to `_savePath` though, so there's nothing to test for writing.

**Comment by twisted-trac** [comment]
|<img alt="racheesingh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|racheesingh commented|
|-|-|

Thanks for the help. Consolidating what I have understood in terms of how we would go about fixing it:

    def __init__(self, savePath):
        ...
        self._savePath = savePath
        ...

+    @property
+    def savePath:
+        return self._savePath

_savePath can be accessed by knownHostsFile.savePath. But knownHostsFile.savePath = "abc" would do result in:
AttributeError: can't set attribute

Does that sound okay?

**Comment by twisted-trac** [comment]
|<img alt="racheesingh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|racheesingh commented|
|-|-|

I apologize for the bad formatting, here is the cleaner looking snippet from the last comment I posted:
```
    def __init__(self, savePath):
        ...
        self._savePath = savePath
        ...

+    @property
+    def savePath:
+        return self._savePath
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

> Does that sound okay?

Please attach changes as diff files, don't paste them as inline comments.  Also please read ReviewProcess and follow the steps outlined there.  Feel free to ask questions about the process - perhaps on the mailing list or the #twisted-dev irc channel - if any part of it isn't clear.

**Comment by twisted-trac** [comment]
|<img alt="racheesingh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|racheesingh removed owner<br>racheesingh set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_13"></a>|@Julian set owner to racheesingh|
|-|-|

This looks OK, but as in your other patch, needs to be tested :).

Take exarkun up on his offer and read that link and come ask questions if you have any.

**Comment by twisted-trac** [comment]
|<img alt="racheesingh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|racheesingh commented|
|-|-|

Thanks :-). As per comment #7, I don't know what test cases I can add in. But I did look at twisted/conch/test/test_knownhosts.py. It has a few statements like:
```
self.assertEqual(exception.path, hostsFile._savePath)
```

Shouldn't these be changed to:
```
self.assertEqual(exception.path, hostsFile.savePath)
```

if the proposed fix was to be accepted?

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_15"></a>|@Julian commented|
|-|-|

Itamar was referring to testing writing the attribute (I.e. testing `h.savePath = 12`), which you don't have to do because it's not a thing that will be supported. In other words, you don't have to write a test here to ensure that a thing that isn't supported isn't supported.

Reading the attribute is though, so at least one test should use the public way to access it.

Changing the existing tests to use your property would sound fine, but I haven't looked at the actual code for the tests specifically yet. I'd recommend doing that and submitting that for review.

**Comment by twisted-trac** [comment]
|<img alt="racheesingh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|racheesingh commented|
|-|-|

Thanks for the feedback, I changed twisted/conch/test/test_knownhosts.py to use hostsFile.savePath instead of hostsFile._savePath. The diff incorporating your comments is attached.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_17"></a>|@Julian commented|
|-|-|

Hey, sorry, this got lost. It's meant to be up for review right?

Putting it up, I'll try to take a look myself tonight, though from a quick glance it looks good.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)|@Julian removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [38492]) Branching to 'knownhostsfile-savepath-6255'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks for your work on this racheesingh.

A few comments:

  1. Methods should be separated from each other by two blank lines
  1. `savePath` should have a docstring
  1. The read-only testing issue here is perhaps a little bit subtle.  If `savePath` were an attribute, then I would say that testing the behavior of assigning to it is not necessary.  Documenting that it should not have its value changed would be sufficient.  However, the patch implements `savePath` as a read-only property.  The property definition is extra code beyond an attribute definition, and it seems like it merits some test coverage.  Fortunately this is a really easy test to write, so it doesn't hurt to err on the side of more testing in this case.
  1. The change should have a news fragment.  This is a feature addition, and it's not a very big one but conch isn't likely to have a lot of feature additions in the next release so we may as well advertise it.

These changes are minor, I'll make them, verify buildbot likes them, and then merge.  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun commented|
|-|-|

(In [38493]) apply patch with fixes

refs [#6255](https://github.com/twisted/twisted/issues/6255)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

Build results - http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/knownhostsfile-savepath-6255

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set status to `closed`|
|-|-|

(In [38494]) Merge knownhostsfile-savepath-6255

Author: racheesingh
Reviewer: Julian, exarkun
Fixes: [#6255](https://github.com/twisted/twisted/issues/6255)

Give `twisted.conch.client.knownhosts.KnownHostsFile` a public `savePath` attribute
which identifies the filesystem path to which information is saved and from which
it is loaded.
